### DISCUSSION TITLE
Controller rules validate break after submit in mode="onBlur"

### DISCUSSION THREAD

**Discussion by lamhungypl** [react-hook-form/react-hook-form#5187] [Q&A] [answered]
I have a custom input and need validation when clicking out, functions worked as expected on first render but after submit, then the form will receive `undefined` for blur validating, it still received correct data for submitting. 
```javascript
<Controller
  render={({ onChange, ...renderProps }) => {
    return (
      <NumberFormat
        {...renderProps}
        onValueChange={value => {
          onChange();
          form.setValue('shippingDays', value.floatValue);
        }}
        decimalScale={0}
        isNumericString
        allowNegative={false}
      />
    );
  }}
  control={form.control}
  name="shippingDays"
  rules={{
    validate: {
      required: data => {
        console.log({ data, mode });
        if (mode === 'send') {
          // not empty and gt 1
          return !toNumber(data) ? `${JSON.stringify(data)}invalidDays` : undefined;
        }
        // not create mode, input is valid by default
        return undefined;
      },
    },
  }}
/>
// native input validate work!!
<input
  name="shippingDaysInput"
  ref={form.register({
    validate: {
      required: data => {
        console.log({ data, mode });
        if (mode === 'send') {
          // not empty and gt 1
          return !toNumber(data) ? `${JSON.stringify(data)}invalidDays` : undefined;
        }
        // not create mode, input is valid by default
        return undefined;
      },
    },
  })}
/>
```
CSB bellow, please notice the Controller field: shipping controller, price, quantity. It looks like it gets the previous value, not the current value of the input

https://user-images.githubusercontent.com/43105576/118180497-cb8fef80-b460-11eb-8889-2fde34d893bc.mov

link: https://codesandbox.io/s/antd-controller-rhf-v7-gdc3v or https://lamhungypl.github.io/RHF-antd-table/

if using version6, (Idk why it needs 2 clicks for setState for switch validate mode, `shippingDays` is optional when edit, greater than 1 when submit -.-)
https://codesandbox.io/s/antd-controller-rhf-v6-mkzef

**Comment by leapful** [comment, marked as answer]
@lamhungypl 

Method "setValue" will not perform validation like "onChange", just setting correct params for "onChange" of react-hook-form will fix the issue.

Working sample: https://codesandbox.io/s/antd-controller-rhf-v7-forked-c0m6f?file=/src/AppTable/AppTable.tsx

**Reply by bluebill1049** [reply]
Thank you again @leapful 🙏

**Reply by lamhungypl** [reply]
I see the issue, thank you. 

I find this params in the docs `onChange(value:any)=>void`, so set `onChange(value.floatValue)` will also fix the broblem.

`any` seems a little bit ...dangerous, how do you find the `{type, target:{name, value}}`?
 
![Screenshot 2021-05-15 at 11 35 29](https://user-images.githubusercontent.com/43105576/118348030-a9d25d80-b571-11eb-810f-eb24dd3033c0.png)

**Reply by leapful** [reply]
@lamhungypl 

Actually, "onChange" accepts both plain value or Event (with field "target" inside), I just copy the handling code another CSB which is using "register", that's why it has an Event object-like :D

It should be plain value in case using "useController" hook or "Controller" component. Sorry for the confusion.
